### DISCUSSION TITLE
Is it possible to run the eye reasoner continously on dynamic data?

### DISCUSSION THREAD

**Discussion by lolgab** [eyereasoner/eye#92] [Q&A]
I want to use eye to materialize triples based on static n3 rules + dynamic triples stored in a triple store.
Is it possible to do that? In case it is, how do you do it?

Considering that the data in the triple store is dynamic, triples can be added and deleted and they influence the results eye gives.
What I wanted is a way to, somehow, run eye in a way that takes into account the diff between the previous input data and the current so it doesn't need to scan the entire data every time a new triple is added or a triple is removed from the triple store.

What I wanted, eventually, was a _smart_ triple store that would answer to SPARQL queries based on the data inside it + the inferred data coming from the n3 rules. But it should be scalable and usable in big applications, a thing that doesn't seem possible if I need to run all the rules on every data change.

**Comment by josd** [comment]
This currently not possible.
What we have been doing is running 1000's of reasoning processes that work together in a start-input-reason-output-stop way.
By creating specialized images of eye (that have precompiled data and rules) the start-input overhead is minimal.
An example of image creation is at https://github.com/eyereasoner/eye/tree/master/reasoning/image

**Comment by lolgab** [comment]
> This is currently not possible.

Is this a future goal of the project or it's completely outside of your goals for eye?

>  What we have been doing is running 1000's of reasoning processes that work together in a start-input-reason-output-stop way.

Imagine an application with a database of million of triples, it is unreasonable to run all the rules on every request that adds/removes data.

>  By creating specialized images of eye (that have precompiled data and rules) the start-input overhead is minimal.

What if the data can change at every interaction with the users, you don't have static data to create the image.

Do you know any strategy for a reasoner like eye to support such scenario? Is it even possible? Do you know any workarounds I can use to make a somewhat incremental reasoning?

**Comment by josd** [comment]
There are no plans to do such things and I propose that you have look at https://2023.eswc-conferences.org/wp-content/uploads/2023/05/paper_Bonte_2023_RoXi.pdf
